test/foo/lib/config/test_configloader.py

```diff
@@ -0,0 +1,26 @@
+import pytest
+
+from foo.lib.config import ConfigLoader, ConfigLoaderException
+
+
+def test_load(monkeypatch):
+    loader = ConfigLoader(defaults={'VAR0': 'default value 0',
+                                    'VAR1': 'default value 1'})
+    monkeypatch.setenv('VAR3', 'not bar')
```

Comment:
General rule of thumb, using system env is a bad practice.

I would expect and order of replacement as follows:
defaults <- env <- config,
The only reason for using config I see, it to quickly get machine settings without producing multiple configurations. It still should be possible to use the config, that can overwrite everything, so there wouldn't be surprises, that it works differently on a given machine.